new menu for armor sorting

akozhevn opened the pull request:
#### Summary

SUMMARY: Interface "Reorganizing armor sorting menu"

#### Purpose of change
Current menu is haphazard with similar (and duplicate) data spread over 4 sections

#### Describe the solution
Move everything into 1 list, so it's easier to read.
Add acid and fire protection.
Add aggregate info for body parts.
Encumberance penalties are now displayed here.
Changed shortcuts to match the rest of the game (u->T, e->W). And added 'd'rop.
Help page is now readable =)

#### Describe alternatives you've considered
Might be good idea to show formulas for calculating penalties and aggregate data.
I would also propose to redirect 'T'ake off menu here, since it's easier to search than in unsorted list

#### Additional context

version 1 (short hand notation):
![armor_sort](https://user-images.githubusercontent.com/6304128/64220902-fd357280-ce98-11e9-9ab7-519802e517d5.png)

version 2 (cut off):
![armor_sort2](https://user-images.githubusercontent.com/6304128/64284769-55678580-cf28-11e9-8d9f-0d2f7db40250.png)

version 3 (full names):
![armor_sort3l](https://user-images.githubusercontent.com/6304128/64364623-4db7e780-cfe1-11e9-91e5-b1b68d643936.png)

version 3 (full names):
![armor_sort3s](https://user-images.githubusercontent.com/6304128/64364641-57414f80-cfe1-11e9-988b-45db487870dd.png)

snipercup commented:
I like the idea. would work nicely on my pc but how will it look on mobile? do we get a scrollbar like with the skills list or will it be multiple pages like advanced inventory?

Leland commented:
Hm, not sure that those extremely abbreviated column names will work well once translated.

(Also a little skeptical of the readability of "strge" and "wrm")

Qrox commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,481 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
-    prot.push_back( name_and_value( space + _( "Environmental:" ),
-                                    string_format( "%3d", static_cast<int>( worn_item.get_env_resist() ) ), width ) );
-    return prot;
+    static std::array< sort_armor_col_data, column_num_entries> col_data = { {
```
Don't use static here, otherwise the translation won't update once loaded even if the language has been changed.

Qrox commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,481 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            {4, _( "cut" ), _( "cut protection." )},
+            {4, _( "acd" ), _( "acid protection." )},
+            {4, _( "fre" ), _( "fire protection." )},
+            {4, _( "env" ), _( "environmental protection." )},
```
Even if you don't abbreviate some of the column headers there's still plenty of space to fit them in the screen. Also, as @Leland said, limiting the character count of a string is bad for translation, and I suggest calculating the column width according to the length of the translated text, instead of limiting the maximum characters a column has.

akozhevn commented:
@snipercup there's a scrollbar, I added small res screenshot

akozhevn commented:
@Qrox I fixed static translations and added printing cut off like this 
`return ( _( name ) ).substr( 0, cut_off );`
There's not that much space in small res (see new screenshot), so have to cut off somewhere.
Spacing matters for displaying data - if data size is same and columns sizes are different it'll be harder to read.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            return width;
+        }
+        std::string get_name() {
+            return ( _( name ) ).substr( 0, cut_off );
```
Please use `utf8_truncate` instead of `substr`. The later will cut right through UTF-8 characters that consist of multiple bytes and thereby produce the wrong output.

Qrox replied:
@akozhevn I see space is tight in small resolution, but this is not really a good way to solve the alignment problem: there's even not a translation comment to tell the translator to limit the character count in their translation, so even if the string is truncated using `utf8_truncate` it will still likely produce some nonsense in other languages.

A solution I can current think of is to split the columns into pages if space is tight (i.e. if the added width of all translated column headers, including padding, exceeds screen width), and allow scrolling the pages using tab and backtab. This way we can fully utilize the screen space, need no heavy abbreviation, and ensure that the data can be viewed in any resolution.

> if data size is same and columns sizes are different it'll be harder to read
You can always take the largest header width as the width for every column.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            {4, 3, "cut",  "cut protection." },
+            {4, 3, "acid",  "acid protection." },
+            {4, 3, "fire",  "fire protection." },
+            {4, 3, "environmental",  "environmental protection." }
```
All of the entries here need translation markers, e.g. `translate_marker`.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
-               this->name == o.name;
-    }
+        int get_col_value( sort_armor_columns col ) {
+            item itm = *itm_iter;
```
```suggestion
            const item &itm = *itm_iter;
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+                        char cond;
+
+                        print_col2 = is_selected ? hilite( c_cyan ) : c_cyan;
+                        cond = ( itm.is_filthy() ? 'f' : ' ' );
```
```suggestion
                        const char cond = ( itm.is_filthy() ? 'f' : ' ' );
```
And remove the definition of `cond` above.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -746,114 +807,85 @@ void player::sort_armor()
+                }
+
+            }
+            help_str += _( string_format(
```
Translating a formatted string does not work. You need to translate the string literal itself.

BevapDin commented on src/player_display.cpp:
```diff
@@ -189,26 +189,26 @@ static std::string dodge_skill_text( double mod )
 }
 
-static int get_encumbrance( const player &p, body_part bp, bool combine )
+static int get_player_encumbrance( const player *p, body_part bp, bool combine )
```
Why change this to take a pointer? It should not be called with a nullptr, so why allow it?

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -746,114 +807,85 @@ void player::sort_armor()
+                if( static_cast<sort_armor_columns>( i ) == column_divider ) {
+                    help_str += "\n";
+                } else {
+                    help_str += string_format( "<color_%s>%s</color> - <color_%s>%s</color>\n",
```
```suggestion
                    help_str += colorize( col_data.get_name(), c_white );
                    help_str += " - ";
                    help_str += colorize( col_data.get_description(), c_light_gray );
                    help_str += "\n";
```
Generally using the `colorize` function is better than adding the color tags manually.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -746,114 +807,85 @@ void player::sort_armor()
-                if( cata::optional<std::list<item>::iterator> new_equip_it =
-                        wear( this->i_at( loc.obtain( *this ) ) ) ) {
+                cata::optional<std::list<item>::iterator> new_equip_it =
+                    wear( this->i_at( loc.obtain( *this ) ) );
```
```suggestion
                    wear( i_at( loc.obtain( *this ) ) );
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -512,222 +618,168 @@ void player::sort_armor()
-        mvwprintz( w_sort_right, point_zero, c_light_gray, _( "(Innermost)" ) );
-        right_print( w_sort_right, 0, 0, c_light_gray, _( "Encumbrance" ) );
+        for( int bp_idx = 0; bp_idx < num_bp; bp_idx++ ) {
+            layering_group_info *l = &items_by_category[bp_idx];
```
```suggestion
            layering_group_info &l = items_by_category[bp_idx];
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -512,222 +618,168 @@ void player::sort_armor()
+            items_by_category[bp].clear();
+        }
+        for( auto elem_it = worn.begin(); elem_it != worn.end(); ++elem_it ) {
+            for( size_t bp = 0; bp < num_bp; bp++ ) {
```
```suggestion
            for( const body_part bp : all_body_parts ) {
```
So you don't have to cast the `size_t` in to a `body_part`.

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+    public:
+        bool is_skipped = false;
+
+        layering_group_info( body_part bp, player *p ) {
```
```suggestion
        layering_group_info( body_part bp, player *p ) : bp( bp ), p( p ) { }
```
And the body is now empty. (It's call member initializer lists.)

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            this->p = p;
+        }
+        void add_item( std::list<item>::iterator itm_i ) {
+            items.push_back( layering_item_info( itm_i, *p, bp ) );
```
```suggestion
            items.emplace_back( itm_i, *p, bp );
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+        }
+        // same for the duration of the menu
+        void set_print_constats( int right_bound, int max_y ) {
+            this->x_bound = right_bound;
```
```suggestion
            x_bound = right_bound;
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
-    wrefresh( w );
-}
+        void set_combined( bool combined ) {
+            this->name = body_part_name_as_heading( all_body_parts[bp], combined ? 2 : 1 );
```
```suggestion
            name = body_part_name_as_heading( all_body_parts[bp], combined ? 2 : 1 );
```
`this` is usually redundant and should not be used in that case (simply because it does not add anything useful).

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+                    case column_storage: {
+                        units::volume storage = itm.get_storage();
+                        mvwprintz( w, point( right_bound, cur_print_y ), print_col, "%*s", col_w,
+                                   ( storage > 0_ml ? format_volume( itm.get_storage() ) : "" ) );
```
```suggestion
                                   storage > 0_ml ? format_volume( itm.get_storage() ) : "" );
```

BevapDin commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            int ret_val;
+            switch( col ) {
+                case column_warmth:
+                    ret_val = itm.get_warmth();
```
```suggestion
                    return itm.get_warmth();
```
Same for all the other placements below. That way you can cut the size of this function in half (get rid of all the `break` statements).

akozhevn commented:
@BevapDin I think I got everything

akozhevn commented:
@Qrox Full names and 2 tabs now. Actually isn't as bad bad as I thought it would be.

akozhevn commented on src/armor_layers.cpp:
```diff
@@ -112,387 +106,528 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
+            }
+
+            //something wrong with using tname, results are too long
+            int elem_w = utf8_width( item::nname( elem.typeId() ) );
```
yeah, that did it

ZhilkinSerg commented:
Feel free to reopen if you want working on it again.
